### DISCUSSION TITLE
DrawText not working(?) when used a second time?

### DISCUSSION THREAD

**Discussion by QuantumToasted** [SixLabors/ImageSharp.Drawing#98] [Q&A]
I don't know if this is a bug or PEBKAC, but I believe it's the latter, but I needed a second pair of eyes because I'm probably just missing something. I did not submit it as an Issue/bug report because I didn't want to make a fool of myself when I likely am just messing something up.

My code is below. Sorry for the consts everywhere. If more context is needed, it can be given.
I've commented the lines that are causing issues for reasons unknown.
<details>
<summary>Code</summary>

```cs
private Image<Rgba32> GenerateDistrict(ICollection<Tribute> tributes, string districtText)
{
    var canvas = new Image<Rgba32>((TRIBUTE_AVATAR_BOUNDS + HORIZONTAL_PAD_MEDIUM) * tributes.Count,
        DISTRICT_HEADER_OFFSET + TRIBUTE_AVATAR_BOUNDS + TRIBUTE_NAME_OFFSET + TRIBUTE_STATUS_OFFSET + SPACE_BETWEEN_DISTRICTS + VERTICAL_PAD * 3);

    canvas.Mutate(x =>
    {
        x.DrawText(new TextGraphicsOptions
                {TextOptions = new TextOptions {HorizontalAlignment = HorizontalAlignment.Center}},
            districtText, Font.CreateFont(DISTRICT_HEADER_FONT_SIZE, FontStyle.Bold), Color.White,
            new PointF(canvas.Width / 2, 0));
    });

    var offset = 0;
    foreach (var tribute in tributes)
    {
        using var box = new Image<Rgba32>(TRIBUTE_AVATAR_BOUNDS,
            DISTRICT_HEADER_OFFSET+ TRIBUTE_AVATAR_BOUNDS + TRIBUTE_NAME_OFFSET + TRIBUTE_STATUS_OFFSET + VERTICAL_PAD * 3);

        using var tributeImage = tribute.GetAvatarImage();

        box.Mutate(x =>
        {
            // Only works for the first tribute. Any tributes after this don't have the text appear.
            x.DrawText(
                new TextGraphicsOptions { TextOptions = new TextOptions { WrapTextWidth = TRIBUTE_AVATAR_BOUNDS, HorizontalAlignment = HorizontalAlignment.Center } },
                tribute.IsAlive ? "Alive" : "Deceased", Font.CreateFont(TRIBUTE_STATUS_FONT_SIZE),
                tribute.IsAlive ? Color.LimeGreen : Color.Red,
                new PointF(offset, DISTRICT_HEADER_OFFSET + VERTICAL_PAD));

            x.DrawImage(tributeImage, new Point(0, DISTRICT_HEADER_OFFSET + TRIBUTE_STATUS_OFFSET + VERTICAL_PAD * 2),
                new GraphicsOptions {AlphaCompositionMode = PixelAlphaCompositionMode.SrcOver});

            // Only works for the first tribute. Any tributes after this don't have the text appear.
            x.DrawText(
                new TextGraphicsOptions {TextOptions = new TextOptions {WrapTextWidth = TRIBUTE_AVATAR_BOUNDS, HorizontalAlignment = HorizontalAlignment.Center}},
                tribute.Name, Font.CreateFont(TRIBUTE_NAME_FONT_SIZE), Color.White,
                new PointF(offset, DISTRICT_HEADER_OFFSET + TRIBUTE_AVATAR_BOUNDS + TRIBUTE_STATUS_OFFSET + VERTICAL_PAD * 3));
        });

        canvas.Mutate(x => x.DrawImage(box, new Point(offset, 0),
            new GraphicsOptions {AlphaCompositionMode = PixelAlphaCompositionMode.SrcOver}));

        offset += HORIZONTAL_PAD_MEDIUM + TRIBUTE_AVATAR_BOUNDS;
    }

    return canvas;
}
```
</details>

Tested with two different maximum numbers of tributes per district.
![image](https://user-images.githubusercontent.com/30324210/97843998-da2a6c80-1caf-11eb-9a07-994c3fa98464.png)
![image](https://user-images.githubusercontent.com/30324210/97844008-deef2080-1caf-11eb-99ed-1c304f4eee3c.png)

Each tribute should have their status (Alive, Deceased) above their avatar, and their name below it. However, only the first tribute (per district) has these. Even though this is processed in a `foreach` loop for each tribute in a completely separate "box", it's going way over my head why this could be the case. Any advice or extra pairs of eyes to find an errant line(s) of code explaining why it's all messed up would be greatly appreciated.

**Comment by tocsoft** [comment]
try removing ```offset += HORIZONTAL_PAD_MEDIUM + TRIBUTE_AVATAR_BOUNDS;``` looks like you are offsetting the text out of the right side of your tribute sub images

**Reply by QuantumToasted** [reply]
> try removing `offset += HORIZONTAL_PAD_MEDIUM + TRIBUTE_AVATAR_BOUNDS;` looks like you are offsetting the text out of the right side of your tribute sub images

I will...try that. `offset` also is responsible for moving the tribute's avatar image as well, wouldn't this just result in the avatar being drawn on top of the first avatar too?

Edit: removed the line. Avatars AND text are now drawn on top of each other.

![image](https://user-images.githubusercontent.com/30324210/97855671-b7548400-1cc0-11eb-8d13-a17219109cbf.png)

The text has magically appeared! 🤔

**Reply by QuantumToasted** [reply]
I figured it out. Removing that line gave me the hint I needed. I don't have a clue why, but I shouldn't have been using `offset` at all in my `.Mutate()` call. Because all I'm doing is making a box that's EXACTLY the size of the status/avatar/name, and then drawing that on a larger canvas. The x values could always be 0 for the boxes, but I only needed `offset` when knowing where to draw the box itself, not the text.

![image](https://user-images.githubusercontent.com/30324210/97857109-dce28d00-1cc2-11eb-953a-f7acbf86be0f.png)

(now, I need to learn how to make text look less awful when it's longer than the wrap length but lacks spaces, but I'll figure that out in time.) Thanks for the help!

**Reply by tocsoft** [reply]
glad you got it sorted
